### PR TITLE
refactor syncer logic more

### PR DESCRIPTION
**Pull request by grosser**
mostly more comments but also logically ordering things more and not relying on modifying the instance vars

/cc @zdrve

### CODE REVIEW

File: lib/kennel/syncer.rb [hunk 1]
```diff
@@ -130,45 +130,56 @@ def calculate_changes
...
-            [e, a]
-          else
-            [nil, a]
-          end
+        # update matching if needed
+        matching.each do |e, a|
+          id = a.fetch(:id)
+          diff = e.diff(a)
+          @update << [id, e, a, diff] if diff.any?
         end
```
Comment (zdrve) [inline lib/kennel/syncer.rb:145]:
```suggestion
        @update = matching.map do |e, a|
          id = a.fetch(:id)
          diff = e.diff(a)
          [id, e, a, diff] if diff.any?
        end.compact
```

Does that work?

File: lib/kennel/syncer.rb [hunk 2]
```diff
@@ -130,45 +130,56 @@ def calculate_changes
...
 
-        # fill details of things we need to compare
-        details = items.map { |e, a| a if e && e.class.api_resource == "dashboard" }.compact
-        @api.fill_details! "dashboard", details
+        # delete previously managed
+        unmatched_actual.each { |a| @delete << [a.fetch(:id), nil, a] if a.fetch(:tracking_id) }
```
Comment (zdrve) [inline lib/kennel/syncer.rb:148]: Likewise,

```suggestion
        @delete = unmatched_actual.map { |a| [a.fetch(:id), nil, a] if a.fetch(:tracking_id) }.compact
```

File: lib/kennel/syncer.rb [hunk 3]
```diff
@@ -130,45 +130,56 @@ def calculate_changes
...
+      matched = []
+      @actual.each do |a|
+        e = matching_expected(a, lookup_map)
+        if e && unmatched_expected.delete?(e)
+          matched << [e, a]
+        else
+          unmatched_actual << a
+        end
+      end.compact
+      [matched, unmatched_expected, unmatched_actual]
```
Comment (zdrve) [inline lib/kennel/syncer.rb:172]: I like this function :-) 

Mild preference for returning a struct so one can't accidentally get the ordering of the triple wrong.

e.g. `Struct.new(:matched, :unmatched_expected, :unmatched_actual).new(matched, unmatched_expected, unmatched_actual`

It's more verbose, but safer.

File: lib/kennel/syncer.rb [hunk 4]
```diff
@@ -130,45 +130,56 @@ def calculate_changes
...
+    end
+
+    # fill details of things we need to compare
+    def fill_details!(details_needed)
+      details_needed = details_needed.map { |e, a| a if e && e.class.api_resource == "dashboard" }.compact
+      @api.fill_details! "dashboard", details_needed
+    end
+
+    def validate_expected_id_not_missing(expected)
```
Comment (zdrve) [inline lib/kennel/syncer.rb:181]: A better name, I approve 👍🏼

**Review by zdrve — APPROVED** [review]

File: lib/kennel/syncer.rb (hunk 1 above)
Comment (grosser) [reply lib/kennel/syncer.rb:145]: ah yes ✨ 
`filter_map` ideally once we are past 2.6

File: lib/kennel/syncer.rb (hunk 3 above)
Comment (grosser) [reply lib/kennel/syncer.rb:172]: just looks so ugly 😭

File: lib/kennel/syncer.rb (hunk 2 above)
Comment (grosser) [reply lib/kennel/syncer.rb:148]: yeah much better now

File: lib/kennel/syncer.rb (hunk 3 above)
Comment (grosser) [reply lib/kennel/syncer.rb:172]: all the code is "expected, actual" so doubt the ordering will go wrong 🤞

**Pull request merged**